daemon/emer-aggregate-tally.c

```diff
@@ -76,113 +86,132 @@ ensure_folder_exists (EmerAggregateTally  *self,
+    g_error ("Failed to execute SQL '%s': (%d) %s", sql, ret, errmsg);
 }
 
-static GVariant*
-load_aggregate_timer_from_path (const char *path)
+static void
+check_sqlite_error (EmerAggregateTally *self,
+                    const char         *tag,
+                    int                 ret)
 {
-  g_autoptr(GMappedFile) mapped_file = NULL;
-  g_autoptr(GVariant) variant = NULL;
-  g_autoptr(GBytes) bytes = NULL;
-  g_autoptr(GError) error = NULL;
+  if (ret == SQLITE_OK || ret == SQLITE_DONE)
+    return;
 
-  mapped_file = g_mapped_file_new (path, FALSE, &error);
-  if (error)
-    {
-      if (!g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        g_warning ("Error loading aggregate timer: %s", error->message);
-      return NULL;
-    }
+  g_error ("%s: (%d) %s",
+           tag,
+           sqlite3_extended_errcode (self->db),
+           sqlite3_errmsg (self->db));
+}
 
-  bytes = g_mapped_file_get_bytes (mapped_file);
-  variant = g_variant_new_from_bytes (G_VARIANT_TYPE (AGGREGATE_TIMER_VARIANT_TYPE),
-                                      bytes,
-                                      FALSE);
+#define CHECK(x) \
+  check_sqlite_error (self, (G_STRLOC), (x))
 
-  if (!g_variant_is_normal_form (variant))
-    {
-      g_autoptr(GError) deletion_error = NULL;
-      g_autoptr(GFile) file = NULL;
+static GVariant *
+text_to_variant (sqlite3_stmt *stmt,
+                 int           i)
+{
+  const gchar *text = (const gchar *)sqlite3_column_text (stmt, i);
+  return g_variant_parse (NULL, text, NULL, NULL, NULL);
+}
 
-      g_warning ("Corrupted aggregate timer at %s, deleting...", path);
+static GVariant *
+column_to_variant (sqlite3_stmt *stmt,
+                   const gchar  *type,
+                   int           i)
+{
+  g_autoptr(GVariant) variant = NULL;
+  gconstpointer sqlite_data;
+  gint size;
 
-      file = g_file_new_for_path (path);
-      g_file_delete (file, NULL, &deletion_error);
-      if (deletion_error)
-        g_warning ("Error deleting corrupted tally file %s: %s",
-                   path, deletion_error->message);
-      return NULL;
-    }
+  sqlite_data = sqlite3_column_blob (stmt, i);
+  if (!sqlite_data)
+    return NULL;
 
-  return g_steal_pointer (&variant);
+  size = sqlite3_column_bytes (stmt, i);
+  gpointer data = g_memdup (sqlite_data, size);
+
+  variant = g_variant_new_from_data (G_VARIANT_TYPE (type),
+                                     data, size, FALSE,
+                                     g_free, data);
+  return swap_bytes_if_big_endian (g_variant_ref_sink (variant));
 }
 
 static guint32
-find_previous_aggregate_timer_counter (EmerAggregateTally *self,
-                                       const char         *aggregate_counter_path)
+column_to_uint32 (sqlite3_stmt *stmt,
+                  int           i)
 {
-  g_autoptr(GVariant) variant = NULL;
-  guint32 previous_counter = 0;
+  int number = sqlite3_column_int (stmt, i);
+  return *((guint32*) &number);
```

Comment:
`CLAMP (number, 0, MAXUINT32)`

Reply:
This is not necessary (and will make the compiler complain about signedness comparison) because the type size doesn't allow anything smaller than 0, nor bigger than, well, the maximum number 1s it can fit.

Reply:
I've made this a bit more readable.